pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DataTypeConversionFunctions.java

```diff
@@ -32,6 +39,33 @@ public class DataTypeConversionFunctions {
   private DataTypeConversionFunctions() {
   }
 
+  @ScalarFunction
+  public static Object cast(Object value, String targetTypeLiteral) {
```

Comment:
ScalarFunction / FunctionInvoker doesn't provide a dynamic return type resolution at runtime --> e.g. determine what's being returned based on the argument passed in. which is exactly what's needed here. 

The SQL here in test are actually compiled into `CastTransformFunction` which does support result metadata --> that's dynamically determined. 

What's missing here is a "CAST" that's not a transform function, which require changes to function invoker.

Reply:
This this works fine when invoked in a filter to handle the case this PR targets - compile time replacement. You're right that it's not getting invoked after aggregation functions though.

Reply:
I can follow up with a separate fix if you agree @richardstartin